# Data Cleaning & Integrity Check for LinkedIn Job Postings (2023) Across 4 Datasets

### In this Jupyter notebook, we delve into the 'LinkedIn Job Postings - 2023', which is composed of 4 distinct datasets. Our mission is to clean and refine each of these datasets, removing any inconsistencies and ensuring their integrity and accuracy. Through meticulous examination and transformation, we aim to amalgamate these datasets into a cohesive and pristine set, primed for subsequent analysis and insights.


In [51]:
# Importing Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [52]:
# Setting Path
path= r'/Users/pierremora/Desktop/LinkedIn Job Postings 2023'

### Importing the first and main database (job_postings)


In [53]:
job_postings = pd.read_csv(r'/Users/pierremora/Desktop/LinkedIn Job Postings 2023/DATA/RAW DATA/job_postings.csv', index_col = False)

### Descriptive Analysis



In [54]:
job_postings.head()

job_id  company_id                     title  \
0   85008768         NaN  Licensed Insurance Agent   
1  133114754  77766802.0             Sales Manager   
2  133196985   1089558.0        Model Risk Auditor   
3  381055942  96654609.0          Business Manager   
4  529257371   1244539.0       NY Studio Assistant   

                                         description  max_salary  med_salary  \
0  While many industries were hurt by the last fe...     52000.0         NaN   
1  Are you a dynamic and creative marketing profe...         NaN         NaN   
2  Join Us as a Model Risk Auditor – Showcase You...         NaN         NaN   
3  Business ManagerFirst Baptist Church ForneyFor...         NaN         NaN   
4  YOU COULD BE ONE OF THE MAGIC MAKERS\nKen Fulk...         NaN         NaN   

   min_salary pay_period formatted_work_type           location  ...  \
0     45760.0     YEARLY           Full-time          Chico, CA  ...   
1         NaN        NaN           Full-time  Santa Clarita, CA  ...   
2         NaN        NaN            Contract       New York, NY  ...   
3         NaN        NaN           Full-time         Forney, TX  ...   
4         NaN        NaN           Full-time       New York, NY  ...   

          expiry  closed_time  formatted_experience_level  skills_desc  \
0  1708301126000          NaN                         NaN          NaN   
1  1695426553000          NaN                         NaN          NaN   
2  1695321872000          NaN                         NaN          NaN   
3  1695424007000          NaN                         NaN          NaN   
4  1708419602000          NaN                         NaN          NaN   

     listed_time posting_domain sponsored  work_type  currency  \
0  1692749193000            NaN         1  FULL_TIME       USD   
1  1692834553000            NaN         0  FULL_TIME       NaN   
2  1692729872000            NaN         0   CONTRACT       NaN   
3  1692832007000            NaN         0  FULL_TIME       NaN   
4  1692867714000            NaN         1  FULL_TIME       NaN   

  compensation_type  
0       BASE_SALARY  
1               NaN  
2               NaN  
3               NaN  
4               NaN  

[5 rows x 27 columns]

### Checking missing values


In [55]:
# counting missing values
missing_values = job_postings.isnull().sum()


In [56]:
# Calculating the percentage of missing values
missing_percentage = (missing_values / len(job_postings)) * 100


In [57]:
# Combining  the results and sort by percentage:
missing_data = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage': missing_percentage
}).sort_values(by='Percentage', ascending=False)


In [58]:
# checking results
print(missing_data[missing_data['Missing Values'] > 0])


                            Missing Values  Percentage
skills_desc                          15742   99.093541
closed_time                          14958   94.158378
med_salary                           14905   93.824751
remote_allowed                       13546   85.270049
max_salary                           10365   65.246129
min_salary                           10365   65.246129
compensation_type                     9384   59.070880
currency                              9384   59.070880
pay_period                            9384   59.070880
applies                               7186   45.234798
posting_domain                        6842   43.069369
application_url                       6091   38.341936
formatted_experience_level            4902   30.857359
views                                 2763   17.392673
company_id                             366    2.303915
description                              1    0.006295


### Dropping Columns with high percentage of missing values (cannot do imputation) 

In [59]:
job_postings.drop(columns=['skills_desc', 'closed_time', 'med_salary'], inplace=True)


### Checking Data types to make decision about imputation in the remote_allowed column

In [60]:
job_postings.dtypes

job_id                          int64
company_id                    float64
title                          object
description                    object
max_salary                    float64
min_salary                    float64
pay_period                     object
formatted_work_type            object
location                       object
applies                       float64
original_listed_time            int64
remote_allowed                float64
views                         float64
job_posting_url                object
application_url                object
application_type               object
expiry                          int64
formatted_experience_level     object
listed_time                     int64
posting_domain                 object
sponsored                       int64
work_type                      object
currency                       object
compensation_type              object
dtype: object

### Changing remote_allowed data type to string to fill missing values as 'not specified'

In [61]:
# filling missing values
job_postings['remote_allowed'].fillna('not_specified', inplace=True)

#Replacing  Numeric Values with String Representations
job_postings['remote_allowed'] = job_postings['remote_allowed'].replace({1.0: 'allowed', 0.0: 'not_allowed'})

print(job_postings['remote_allowed'].value_counts())



remote_allowed
not_specified    13546
allowed           2340
Name: count, dtype: int64


### Eliminating unnecessary columns  for analysis





In [62]:
#dropping columns
job_postings.drop(columns=['job_posting_url','application_url','posting_domain'], inplace=True)


In [21]:
#checking actual columns 
job_postings.dtypes

job_id                          int64
company_id                    float64
title                          object
description                    object
max_salary                    float64
min_salary                    float64
pay_period                     object
formatted_work_type            object
location                       object
applies                       float64
original_listed_time            int64
remote_allowed                 object
views                         float64
application_type               object
expiry                          int64
formatted_experience_level     object
listed_time                     int64
sponsored                       int64
work_type                      object
currency                       object
compensation_type              object
dtype: object

In [63]:
job_postings.head(5)

job_id  company_id                     title  \
0   85008768         NaN  Licensed Insurance Agent   
1  133114754  77766802.0             Sales Manager   
2  133196985   1089558.0        Model Risk Auditor   
3  381055942  96654609.0          Business Manager   
4  529257371   1244539.0       NY Studio Assistant   

                                         description  max_salary  min_salary  \
0  While many industries were hurt by the last fe...     52000.0     45760.0   
1  Are you a dynamic and creative marketing profe...         NaN         NaN   
2  Join Us as a Model Risk Auditor – Showcase You...         NaN         NaN   
3  Business ManagerFirst Baptist Church ForneyFor...         NaN         NaN   
4  YOU COULD BE ONE OF THE MAGIC MAKERS\nKen Fulk...         NaN         NaN   

  pay_period formatted_work_type           location  applies  ...  \
0     YEARLY           Full-time          Chico, CA      NaN  ...   
1        NaN           Full-time  Santa Clarita, CA      NaN  ...   
2        NaN            Contract       New York, NY      1.0  ...   
3        NaN           Full-time         Forney, TX      NaN  ...   
4        NaN           Full-time       New York, NY      NaN  ...   

   remote_allowed views    application_type         expiry  \
0   not_specified   5.0  ComplexOnsiteApply  1708301126000   
1   not_specified   NaN  ComplexOnsiteApply  1695426553000   
2   not_specified  17.0  ComplexOnsiteApply  1695321872000   
3   not_specified   NaN  ComplexOnsiteApply  1695424007000   
4   not_specified   2.0  ComplexOnsiteApply  1708419602000   

   formatted_experience_level    listed_time  sponsored  work_type currency  \
0                         NaN  1692749193000          1  FULL_TIME      USD   
1                         NaN  1692834553000          0  FULL_TIME      NaN   
2                         NaN  1692729872000          0   CONTRACT      NaN   
3                         NaN  1692832007000          0  FULL_TIME      NaN   
4                         NaN  1692867714000          1  FULL_TIME      NaN   

  compensation_type  
0       BASE_SALARY  
1               NaN  
2               NaN  
3               NaN  
4               NaN  

[5 rows x 21 columns]

### Checking for fully duplicates

In [64]:
# Identifying fully duplicated rows
full_duplicates = job_postings[job_postings.duplicated()]

# Counting the number of fully duplicated rows
num_full_duplicates = len(full_duplicates)
print(f"Number of fully duplicated rows: {num_full_duplicates}")


Number of fully duplicated rows: 0


In [65]:
## checking for duplicates in KEY columns
key_columns = ['title', 'description', 'company_id']
duplicates_by_key = job_postings[job_postings.duplicated(subset=key_columns)]

# Counting the number of duplicates based on key columns
num_key_duplicates = len(duplicates_by_key)
print(f"Number of duplicates based on key columns: {num_key_duplicates}")

Number of duplicates based on key columns: 1789


In [66]:
### Analysis on duplicates based on key columns
print(duplicates_by_key.head(10)) 



         job_id  company_id  \
16   2148434019      1016.0   
28   2148434726     13074.0   
254  3690888827     11753.0   
301  3692361910      3671.0   
302  3692361914      3671.0   
304  3692362852      3671.0   
305  3692362854      3671.0   
309  3692363702      3671.0   
310  3692363706      3671.0   
311  3692363707      3671.0   

                                                 title  \
16   Virtual Sales Associate Account Manager, Women...   
28                          Mental Health Professional   
254                                   Personal Trainer   
301            Block Advisor - Remote Tax Professional   
302            Block Advisor - Remote Tax Professional   
304            Block Advisor - Remote Tax Professional   
305            Block Advisor - Remote Tax Professional   
309            Block Advisor - Remote Tax Professional   
310            Block Advisor - Remote Tax Professional   
311            Block Advisor - Remote Tax Professional   

                   

### After analysing the duplicated based on key columns, there is evidence that the job posts are different because each offer is for different city in the United States


## Data Type checking

### Some columns data type is 'float64', to avoid floating numbers, it will be changed to 'Int64', keeping the missing values (NaN)

In [67]:
# List of columns to convert to Int64
columns_to_convert_to_Int64 = ['applies', 'company_id', 'views', 'max_salary', 'min_salary']

# Convert specified columns to Int64 using pd.Series
for col in columns_to_convert_to_Int64:
    job_postings[col] = pd.Series(job_postings[col], dtype="Int64")

# Checking the data types of the specified columns to confirm the conversion
print(job_postings[columns_to_convert_to_Int64].dtypes)





applies         Int64
company_id      Int64
views           Int64
max_salary    float64
min_salary    float64
dtype: object


### Data Accuracy

Checking data accuracy for numerical columns involves ensuring that the values in these columns make logical sense and are within expected ranges.

In [68]:
#checking negative values

incorrect_salaries = job_postings[job_postings['company_id'] < 0]

print(incorrect_salaries)


Empty DataFrame
Columns: [job_id, company_id, title, description, max_salary, min_salary, pay_period, formatted_work_type, location, applies, original_listed_time, remote_allowed, views, application_type, expiry, formatted_experience_level, listed_time, sponsored, work_type, currency, compensation_type]
Index: []

[0 rows x 21 columns]


In [69]:
#checking negative salaries
incorrect_salaries_max = job_postings[job_postings['max_salary'] < 0]

print(incorrect_salaries)

Empty DataFrame
Columns: [job_id, company_id, title, description, max_salary, min_salary, pay_period, formatted_work_type, location, applies, original_listed_time, remote_allowed, views, application_type, expiry, formatted_experience_level, listed_time, sponsored, work_type, currency, compensation_type]
Index: []

[0 rows x 21 columns]


No negative Values found

### Data Integrity & Consistency
Ensure relationships between columns are maintained.

In [70]:

# Counting the number of job applications based on each location
location_counts = job_postings['location'].value_counts()

print(location_counts)



location
United States     1133
New York, NY       398
Chicago, IL        267
Houston, TX        243
Atlanta, GA        207
                  ... 
Visalia, CA          1
Page, AZ             1
Blacksburg, VA       1
Windsor, WI          1
Vicksburg, MS        1
Name: count, Length: 3010, dtype: int64


In [71]:
# Counting the number of commas in each entry of the 'location' column
location_format_counts = job_postings['location'].str.count(',').value_counts()

print(location_format_counts)


location
1    13939
0     1891
2       56
Name: count, dtype: int64


13,939 entries have one comma, suggesting a "City, Country/State" format.

1,891 entries don't have any commas, which likely indicates just a country or a standalone city name.

56 entries have two commas, suggesting a more detailed format like "City, State, Country" or variations.

In [72]:
# Extract 20 examples of entries without a comma
additional_no_comma_examples = job_postings[job_postings['location'].str.count(',') == 0]['location'].sample(20)

print(additional_no_comma_examples)


6370                             United States
4270                             United States
6344          Salt Lake City Metropolitan Area
14758                 Greater Hattiesburg Area
843                Dallas-Fort Worth Metroplex
4111                             United States
7437          Salt Lake City Metropolitan Area
15100          New York City Metropolitan Area
4323         Greater Minneapolis-St. Paul Area
6243                             United States
891                              United States
6440     Harrisonburg-Staunton-Waynesboro Area
109                              United States
4257                    San Francisco Bay Area
8529                             United States
1808                             United States
1860                             United States
13143                     Greater Chicago Area
7630                 Greater Pittsburgh Region
5634              Washington DC-Baltimore Area
Name: location, dtype: object


These entries consist of country names, broader metropolitan areas, specific regions, and some potentially ambiguous entries like "NAMER".

In [73]:
# Dropping rows where the location column doesn't contain commas
job_postings = job_postings[job_postings['location'].str.count(',') > 0]


In [74]:
# Check the unique counts of commas in the location column
remaining_comma_counts = job_postings['location'].str.count(',').value_counts()

print(remaining_comma_counts)



location
1    13939
2       56
Name: count, dtype: int64


In [76]:
# Dropping rows where the location column doesn't contain commas
job_postings = job_postings[job_postings['location'].str.count(',') == 1]


In [77]:
# Check the unique counts of commas in the location column
remaining_comma_counts = job_postings['location'].str.count(',').value_counts()

print(remaining_comma_counts)

location
1    13939
Name: count, dtype: int64


### After dropping rows where the location column doesn't contain commas, the dataset now has 13,939 entries. This ensures that the remaining entries have a consistent "City, Country/State" format or variations of it, which will improve the consistency of the location data.

In [78]:
# Split the location column to separate city from state/country
job_postings['city'] = job_postings['location'].str.split(', ').str[0]
job_postings['state_or_country'] = job_postings['location'].str.split(', ').str[1]


### This code uses the str.split() method to split the location column at the comma, and then assigns the resulting parts to the respective new columns

In [79]:
#checking results
job_postings.head(2)

job_id  company_id                     title  \
0   85008768        <NA>  Licensed Insurance Agent   
1  133114754    77766802             Sales Manager   

                                         description  max_salary  min_salary  \
0  While many industries were hurt by the last fe...     52000.0     45760.0   
1  Are you a dynamic and creative marketing profe...         NaN         NaN   

  pay_period formatted_work_type           location  applies  ...  \
0     YEARLY           Full-time          Chico, CA     <NA>  ...   
1        NaN           Full-time  Santa Clarita, CA     <NA>  ...   

     application_type         expiry  formatted_experience_level  \
0  ComplexOnsiteApply  1708301126000                         NaN   
1  ComplexOnsiteApply  1695426553000                         NaN   

     listed_time  sponsored  work_type  currency  compensation_type  \
0  1692749193000          1  FULL_TIME       USD        BASE_SALARY   
1  1692834553000          0  FULL_TIME       NaN                NaN   

            city state_or_country  
0          Chico               CA  
1  Santa Clarita               CA  

[2 rows x 23 columns]

In [80]:
# Renaming the columns
job_postings = job_postings.rename(columns={
    'city': 'job_location_city',
    'state_or_country': 'job_location_state'
})


In [81]:
# dropping location column
job_postings = job_postings.drop(columns=['location'])


In [82]:
#creating a dictionary of US states to replace abbreviations to full state names
# Dictionary mapping state abbreviations to full names
state_abbreviations = {
    'AL': 'Alabama', 'AK': 'Alaska', 'AZ': 'Arizona', 'AR': 'Arkansas', 'CA': 'California',
    'CO': 'Colorado', 'CT': 'Connecticut', 'DE': 'Delaware', 'FL': 'Florida', 'GA': 'Georgia',
    'HI': 'Hawaii', 'ID': 'Idaho', 'IL': 'Illinois', 'IN': 'Indiana', 'IA': 'Iowa',
    'KS': 'Kansas', 'KY': 'Kentucky', 'LA': 'Louisiana', 'ME': 'Maine', 'MD': 'Maryland',
    'MA': 'Massachusetts', 'MI': 'Michigan', 'MN': 'Minnesota', 'MS': 'Mississippi', 'MO': 'Missouri',
    'MT': 'Montana', 'NE': 'Nebraska', 'NV': 'Nevada', 'NH': 'New Hampshire', 'NJ': 'New Jersey',
    'NM': 'New Mexico', 'NY': 'New York', 'NC': 'North Carolina', 'ND': 'North Dakota', 'OH': 'Ohio',
    'OK': 'Oklahoma', 'OR': 'Oregon', 'PA': 'Pennsylvania', 'RI': 'Rhode Island', 'SC': 'South Carolina',
    'SD': 'South Dakota', 'TN': 'Tennessee', 'TX': 'Texas', 'UT': 'Utah', 'VT': 'Vermont',
    'VA': 'Virginia', 'WA': 'Washington', 'WV': 'West Virginia', 'WI': 'Wisconsin', 'WY': 'Wyoming',
    'DC': 'District of Columbia', 'PR': 'Puerto Rico'
}

# Replacing state abbreviations with full names
job_postings['job_location_state'] = job_postings['job_location_state'].map(state_abbreviations).fillna(job_postings['job_location_state'])


### This code will replace the state abbreviations in the job_location_state column with their full names. If an abbreviation isn't found in the dictionary, the original value will remain unchanged.

In [84]:
# counting states
state_counts = job_postings['job_location_state'].value_counts()

for state, count in state_counts.items():
    print(f"{state}: {count}")


California: 1825
Texas: 1305
New York: 783
Florida: 757
United States: 661
Illinois: 565
Georgia: 452
North Carolina: 449
Pennsylvania: 441
Massachusetts: 438
Virginia: 436
Washington: 408
Ohio: 397
New Jersey: 386
Michigan: 329
Arizona: 311
Maryland: 310
Colorado: 287
Minnesota: 286
Tennessee: 259
South Carolina: 233
Wisconsin: 230
Missouri: 210
Indiana: 186
Connecticut: 145
Utah: 141
Oregon: 127
Nevada: 124
Oklahoma: 111
Iowa: 110
District of Columbia: 109
Kansas: 98
Louisiana: 97
Alabama: 95
Kentucky: 92
Idaho: 68
Arkansas: 66
Delaware: 56
Nebraska: 54
New Mexico: 49
Mississippi: 45
Texas Metropolitan Area: 44
Rhode Island: 40
New Hampshire: 35
West Virginia: 34
Hawaii: 33
Alaska: 31
North Dakota: 29
South Dakota: 20
Montana: 20
Oregon Metropolitan Area: 18
Wyoming: 18
Maine: 15
Ohio Metropolitan Area: 15
Vermont: 13
New York Metropolitan Area: 10
South Carolina Metropolitan Area: 8
South Carolina Area: 3
Wisconsin Metropolitan Area: 3
Indiana Metropolitan Area: 3
AZ Area: 3
Alabama

In [85]:
# Filter rows to keep only those with states found in the state_abbreviations dictionary
job_postings = job_postings[job_postings['job_location_state'].isin(state_abbreviations.values())]


By applying this filter, the dataset will only include rows where the job_location_state column contains recognized U.S. state names.

In [86]:
# recheck 
# counting states
state_counts = job_postings['job_location_state'].value_counts()

for state, count in state_counts.items():
    print(f"{state}: {count}")

California: 1825
Texas: 1305
New York: 783
Florida: 757
Illinois: 565
Georgia: 452
North Carolina: 449
Pennsylvania: 441
Massachusetts: 438
Virginia: 436
Washington: 408
Ohio: 397
New Jersey: 386
Michigan: 329
Arizona: 311
Maryland: 310
Colorado: 287
Minnesota: 286
Tennessee: 259
South Carolina: 233
Wisconsin: 230
Missouri: 210
Indiana: 186
Connecticut: 145
Utah: 141
Oregon: 127
Nevada: 124
Oklahoma: 111
Iowa: 110
District of Columbia: 109
Kansas: 98
Louisiana: 97
Alabama: 95
Kentucky: 92
Idaho: 68
Arkansas: 66
Delaware: 56
Nebraska: 54
New Mexico: 49
Mississippi: 45
Rhode Island: 40
New Hampshire: 35
West Virginia: 34
Hawaii: 33
Alaska: 31
North Dakota: 29
Montana: 20
South Dakota: 20
Wyoming: 18
Maine: 15
Vermont: 13


In [88]:
city_counts = job_postings['job_location_city'].value_counts()
print(city_counts)


job_location_city
New York              398
Chicago               267
Houston               243
Atlanta               207
Los Angeles           188
                     ... 
Bloomington–Normal      1
Screven                 1
Charles City            1
Williamson County       1
Vicksburg               1
Name: count, Length: 2443, dtype: int64


In [93]:
# exporting dataframe to pickle
# Export data to pkl

job_postings.to_csv(os.path.join(path, 'DATA','CLEAN DATA', 'job_postings_clean.csv'))

Creating a data profile table

In [94]:

# Define the profiling data
data = {
    'Variables': ['job_id', 'company_id', 'title', 'description', 'max_salary', 'min_salary', 'pay_period', 'formatted_work_type', 'job_location_city', 
                  'applies', 'original_listed_time', 'remote_allowed', 'views', 'application_type', 'formatted_experience_level', 'listed_time', 
                  'sponsored', 'work_type', 'currency', 'compensation_type', 'job_location_city', 'job_location_state'],
    'Description': ['unique id for job', 'unique id for company', 'Job title', 'Job description', 'Maximum salary', 'Minimum salary', 'Pay period for salary',
                    'Type of work', 'location of job offer', 'Number of applications submitted', 'Original time the job was listed', 'Whether job permits remote work',
                    'Number of times the job posting viewed', 'Type of application process', 'Job experience level', 'Time when the job was listed', 
                    'Whether the job listing is sponsored', 'Type of work associated with the job', 'Currency of the salary', 'Type of compensation for the job', 
                    'location of job offer', 'state of the job offer'],
    'Data Types': ['int64', 'float64', 'object', 'object', 'float64', 'float64', 'object', 'object', 'object', 
                   'float64', 'int64', 'object', 'float64', 'object', 'object', 'int64', 
                   'int64', 'object', 'object', 'object', 'object', 'object'],
    'Time-variant/-invariant': ['Invariant', 'Invariant', 'Invariant', 'Invariant', 'Invariant', 'Invariant', 'Invariant', 'Invariant', 'Invariant',
                                'Variant', 'Invariant', 'Invariant', 'Variant', 'Invariant', 'Invariant', 'Invariant',
                                'Invariant', 'Invariant', 'Invariant', 'Invariant', 'Invariant', 'Invariant'],
    'Structured/Unstructured': ['Structured', 'Structured', 'Structured', 'Unstructured', 'Structured', 'Structured', 'Structured', 'Structured', 'Structured',
                                'Structured', 'Structured', 'Structured', 'Structured', 'Structured', 'Structured', 'Structured',
                                'Structured', 'Structured', 'Structured', 'Structured', 'Structured', 'Structured'],
    'Qualitative/Quantitative': ['Quantitative', 'Quantitative', 'Qualitative', 'Qualitative', 'Quantitative', 'Quantitative', 'Qualitative', 'Qualitative', 'Qualitative',
                                 'Quantitative', 'Quantitative', 'Qualitative', 'Quantitative', 'Qualitative', 'Qualitative', 'Quantitative',
                                 'Quantitative', 'Qualitative', 'Qualitative', 'Qualitative', 'Qualitative', 'Qualitative'],
    'Qualitative: Nominal/Ordinal': ['-', '-', 'Nominal', 'Nominal', '-', '-', 'Nominal', 'Nominal', 'Nominal',
                                     '-', '-', 'Nominal', '-', 'Nominal', 'Ordinal', '-',
                                     '-', 'Nominal', 'Nominal', 'Nominal', 'Nominal', 'Nominal'],
    'Quantitative: Discrete/Continuous': ['Discrete', 'Discrete', '-', '-', 'Continuous', 'Continuous', '-', '-', '-',
                                          'Discrete', 'Discrete', '-', 'Discrete', '-', '-', 'Discrete',
                                          'Discrete', '-', '-', '-', '-', '-']
}

# Create DataFrame
job_postings_profiling = pd.DataFrame(data)

job_postings_profiling


Variables                             Description  \
0                       job_id                       unique id for job   
1                   company_id                   unique id for company   
2                        title                               Job title   
3                  description                         Job description   
4                   max_salary                          Maximum salary   
5                   min_salary                          Minimum salary   
6                   pay_period                   Pay period for salary   
7          formatted_work_type                            Type of work   
8            job_location_city                   location of job offer   
9                      applies        Number of applications submitted   
10        original_listed_time        Original time the job was listed   
11              remote_allowed         Whether job permits remote work   
12                       views  Number of times the job posting viewed   
13            application_type             Type of application process   
14  formatted_experience_level                    Job experience level   
15                 listed_time            Time when the job was listed   
16                   sponsored    Whether the job listing is sponsored   
17                   work_type    Type of work associated with the job   
18                    currency                  Currency of the salary   
19           compensation_type        Type of compensation for the job   
20           job_location_city                   location of job offer   
21          job_location_state                  state of the job offer   

   Data Types Time-variant/-invariant Structured/Unstructured  \
0       int64               Invariant              Structured   
1     float64               Invariant              Structured   
2      object               Invariant              Structured   
3      object               Invariant            Unstructured   
4     float64               Invariant              Structured   
5     float64               Invariant              Structured   
6      object               Invariant              Structured   
7      object               Invariant              Structured   
8      object               Invariant              Structured   
9     float64                 Variant              Structured   
10      int64               Invariant              Structured   
11     object               Invariant              Structured   
12    float64                 Variant              Structured   
13     object               Invariant              Structured   
14     object               Invariant              Structured   
15      int64               Invariant              Structured   
16      int64               Invariant              Structured   
17     object               Invariant              Structured   
18     object               Invariant              Structured   
19     object               Invariant              Structured   
20     object               Invariant              Structured   
21     object               Invariant              Structured   

   Qualitative/Quantitative Qualitative: Nominal/Ordinal  \
0              Quantitative                            -   
1              Quantitative                            -   
2               Qualitative                      Nominal   
3               Qualitative                      Nominal   
4              Quantitative                            -   
5              Quantitative                            -   
6               Qualitative                      Nominal   
7               Qualitative                      Nominal   
8               Qualitative                      Nominal   
9              Quantitative                            -   
10             Quantitative                            -   
11              Qualitative                      Nominal   
12             Quantitative               

In [95]:
# exporting dataframe to cvs


job_postings_profiling.to_csv(os.path.join(path, 'DATA','Data profile', 'job_postings_data_profile.csv'))

In [97]:
job_postings.describe()

job_id       company_id      max_salary     min_salary  \
count  1.315800e+04          12880.0     4469.000000    4469.000000   
mean   3.691411e+09   9450037.235637    83540.027984   59017.067548   
std    1.048243e+08  21466476.753983    86710.403670   58370.833324   
min    8.500877e+07           1009.0       10.000000      10.000000   
25%    3.693073e+09          11038.5       40.000000      30.250000   
50%    3.697363e+09         165464.0    75000.000000   58240.000000   
75%    3.699413e+09        4822178.0   133000.000000   93400.000000   
max    3.701374e+09       98556556.0  1000000.000000  750000.000000   

         applies  original_listed_time      views        expiry   listed_time  \
count     6833.0          1.315800e+04    10636.0  1.315800e+04  1.315800e+04   
mean   15.370408          1.692633e+12  51.932023  1.696238e+12  1.692767e+12   
std    31.427274          4.667735e+08  93.049319  3.208041e+09  1.075535e+08   
min          1.0          1.686211e+12        1.0  1.694271e+12  1.691672e+12   
25%          2.0          1.692731e+12        5.0  1.695328e+12  1.692736e+12   
50%          5.0          1.692737e+12       20.0  1.695338e+12  1.692745e+12   
75%         15.0          1.692824e+12       58.0  1.695425e+12  1.692831e+12   
max        832.0          1.692868e+12     1549.0  1.708420e+12  1.692868e+12   

          sponsored  
count  13158.000000  
mean       0.305442  
std        0.460612  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max        1.000000

### Importing second dataset (company_industries)


In [98]:
company_industries = pd.read_csv(r'/Users/pierremora/Desktop/LinkedIn Job Postings 2023/DATA/RAW DATA/company_industries.csv', index_col = False)

### Descpritive analysis of the dataset

In [102]:
#checking data types
company_industries.dtypes

company_id     int64
industry      object
dtype: object

In [103]:
#checking content
company_industries.describe()

company_id
count  1.588000e+04
mean   1.072469e+07
std    2.274282e+07
min    1.009000e+03
25%    1.338600e+04
50%    3.474100e+05
75%    9.308035e+06
max    9.856222e+07

In [104]:
company_industries.head(5)

company_id                               industry
0    81149246                       Higher Education
1    10033339      Information Technology & Services
2     6049228                             Accounting
3     2641066  Electrical & Electronic Manufacturing
4    96649998                Marketing & Advertising

### Checking missing values

In [106]:
# counting missing values
missing_values3 = company_industries.isnull().sum()
# Calculating the percentage of missing values
missing_percentage3 = (missing_values3 / len(company_industries)) * 100
# Combining  the results and sort by percentage:
missing_data3 = pd.DataFrame({
    'Missing Values': missing_values3,
    'Percentage': missing_percentage3
}).sort_values(by='Percentage', ascending=False)
# checking results
print(missing_data3[missing_data3['Missing Values'] > 0])


Empty DataFrame
Columns: [Missing Values, Percentage]
Index: []


### Checking fully duplicates

In [107]:
# Identifying fully duplicated rows
full_duplicates3 = company_industries[company_industries.duplicated()]

# Counting the number of fully duplicated rows
num_full_duplicates3 = len(full_duplicates3)
print(f"Number of fully duplicated rows: {num_full_duplicates3}")

Number of fully duplicated rows: 9877


In [109]:
company_industries.shape

(15880, 2)

In [110]:
job_postings.shape

(13158, 22)

In [111]:
# Removing fully duplicated rows based on both company_id and industry
company_industries_no_dup = company_industries.drop_duplicates(subset=['company_id', 'industry'], keep='first')




In [112]:
company_industries.shape

(15880, 2)

In [113]:
# counting company_id
company_counts = company_industries_no_dup['company_id'].value_counts()

for company, count in company_counts.items():
    print(f"{company}: {count}")


81149246: 1
1576969: 1
14780375: 1
3131483: 1
162749: 1
68177267: 1
29331: 1
33282143: 1
51716: 1
79696510: 1
56861: 1
64883: 1
1228217: 1
2671586: 1
110372: 1
4918: 1
52143761: 1
686294: 1
497017: 1
18209861: 1
15564: 1
25017205: 1
53491671: 1
732744: 1
49644642: 1
13637042: 1
8455: 1
2690: 1
24787197: 1
162724: 1
14133: 1
2087: 1
111485: 1
1618718: 1
298190: 1
65294553: 1
1115776: 1
89877420: 1
77049371: 1
28439179: 1
35498950: 1
6104: 1
162912: 1
77301: 1
3196: 1
24122185: 1
1916708: 1
1162908: 1
2681277: 1
2480405: 1
6702: 1
26067: 1
217338: 1
35566147: 1
165422: 1
163852: 1
18401: 1
3812750: 1
367977: 1
665858: 1
4030721: 1
28810: 1
911928: 1
10877848: 1
143720: 1
814025: 1
227878: 1
164134: 1
40692536: 1
1511: 1
12250: 1
8151492: 1
27185258: 1
851980: 1
26630: 1
6470711: 1
2528: 1
2708254: 1
88016: 1
493453: 1
82315716: 1
71910: 1
15656: 1
1627: 1
17998520: 1
5064123: 1
41147: 1
84418: 1
6367266: 1
1526: 1
3260: 1
368844: 1
15223686: 1
78310268: 1
157346: 1
64981: 1
10447496: 1
8

In [114]:
# counting industries
industry_counts = company_industries_no_dup['industry'].value_counts()

for industry, count in industry_counts.items():
    print(f"{industry}: {count}")

Staffing & Recruiting: 780
Information Technology & Services: 691
Hospital & Health Care: 426
Computer Software: 271
Financial Services: 231
Construction: 154
Retail: 152
Non-profit Organization Management: 129
Real Estate: 126
Management Consulting: 117
Higher Education: 106
Marketing & Advertising: 101
Insurance: 100
Transportation/Trucking/Railroad: 97
Law Practice: 92
Government Administration: 84
Hospitality: 81
Automotive: 77
Consumer Goods: 70
Biotechnology: 67
Apparel & Fashion: 65
Food & Beverages: 61
Electrical & Electronic Manufacturing: 60
Medical Device: 60
Health, Wellness & Fitness: 59
Banking: 57
Mechanical Or Industrial Engineering: 56
Mental Health Care: 54
Oil & Energy: 51
Education Management: 51
Pharmaceuticals: 49
Utilities: 49
Human Resources: 45
Entertainment: 45
Machinery: 42
Building Materials: 42
Food Production: 42
Accounting: 41
Environmental Services: 41
Telecommunications: 40
Research: 39
Defense & Space: 38
Civil Engineering: 36
Chemicals: 32
Aviation & 

### Checking Data Accuracy

In [117]:
# Checking for invalid company_id values (non-positive values or non-integer format)
invalid_company_ids = company_industries_no_dup[~company_industries_no_dup['company_id'].apply(lambda x: str(x).isdigit() and x > 0)]

invalid_company_ids_count = len(invalid_company_ids)


I used the apply method on the company_id column to check each value.

The lambda function lambda x: str(x).isdigit() and x > 0 checks two conditions:

str(x).isdigit(): This ensures that the value, when converted to a string, consists only of digits, which confirms it's an integer.

x > 0: This ensures that the value is positive.

The ~ operator is used to filter rows that do NOT satisfy the conditions, i.e., invalid company_id values.

Finally, I counted the number of such invalid values using len.

In [118]:
# exporting dataframe to cvs


company_industries_no_dup.to_csv(os.path.join(path, 'DATA','CLEAN DATA', 'company_industries_clean.csv'))

## Importing the third dataset (companies)

In [119]:
companies_names = pd.read_csv(r'/Users/pierremora/Desktop/LinkedIn Job Postings 2023/DATA/RAW DATA/companies.csv', index_col = False)

In [120]:
#checking data types
companies_names.dtypes

company_id        int64
name             object
description      object
company_size    float64
state            object
country          object
city             object
zip_code         object
address          object
url              object
dtype: object

In [122]:
#analyzing database
companies_names.describe()

company_id  company_size
count  6.063000e+03   5474.000000
mean   6.669791e+07      3.764706
std    4.276660e+08      2.057414
min    1.009000e+03      1.000000
25%    9.488450e+04      2.000000
50%    2.174958e+06      4.000000
75%    1.834672e+07      5.000000
max    3.700153e+09      7.000000

### Checking missing Values

In [123]:
# Calculating the number of missing values for each column
missing_values4 = companies_names.isnull().sum()

# Calculating the percentage of missing values for each column
missing_percentages4 = (missing_values4 / len(companies_names)) * 100

# Creating a DataFrame to display the results
missing_data4 = pd.DataFrame({'Missing Values': missing_values4, 'Percentage (%)': missing_percentages4})

# Sort the results by the percentage of missing values in descending order
missing_data4 = missing_data4.sort_values(by='Percentage (%)', ascending=False)


In [124]:
# Printing the results
print(missing_data4)

              Missing Values  Percentage (%)
company_size             589        9.714663
description               64        1.055583
zip_code                   8        0.131948
address                    5        0.082467
state                      3        0.049480
company_id                 0        0.000000
name                       0        0.000000
country                    0        0.000000
city                       0        0.000000
url                        0        0.000000


Due to low percentage of missing values, these values will remain empty in the dataset

### Checking for fully duplicates

In [126]:
# Checking for fully duplicated rows
fully_duplicated_rows4 = companies_names[companies_names.duplicated(keep=False)]

# Counting the number of fully duplicated rows
fully_duplicated_count4 = len(fully_duplicated_rows4)


In [127]:
# Printing the results
print(f"There are {fully_duplicated_count4} fully duplicated rows in the dataset.")


There are 0 fully duplicated rows in the dataset.


There are no fully duplicated rows in the companies.csv dataset. This means every row in the dataset is unique.

### data consistency

In [128]:
# Counting the number of unique values in company_id
unique_company_ids2 = companies_names['company_id'].nunique()

# Getting the total number of rows in the dataset
total_rows = len(companies_names)

# Check if the number of unique company_ids is equal to the total number of rows
is_unique = unique_company_ids2 == total_rows


In this code:

I first counted the number of unique values in the company_id column.

Then, I determined the total number of rows in the dataset.

Finally, I compared the two numbers to see if they were equal. If they are equal, it indicates that the company_id is unique for each row.

In [129]:
# Printing the results
if is_unique:
    print("The company_id column is unique for each row.")
else:
    print("The company_id column is not unique. There are duplicate values.")

The company_id column is unique for each row.


In [130]:
# Counting occurrences of each country
country_counts2 = companies_names['country'].value_counts()



In [131]:
#printing results
print(country_counts2)

country
US    5419
0      156
GB     129
CA      71
IN      39
DE      35
CH      29
AU      19
FR      19
NL      15
SE      11
CN      11
FI       9
SG       8
DK       8
IE       7
IT       7
OO       6
JP       5
ES       5
IL       4
MX       4
AT       4
KR       3
AE       3
BE       3
BM       3
HK       3
TW       3
NZ       2
SA       2
AR       2
NO       2
HU       1
UY       1
ID       1
VN       1
KY       1
HR       1
TR       1
BR       1
CZ       1
GE       1
MD       1
IS       1
PR       1
CO       1
KE       1
LU       1
PH       1
Name: count, dtype: int64


In [132]:

# Mapping of ISO codes to full country names
iso_to_full_country = {
    'US': 'United States',
    'GB': 'United Kingdom',
    'CA': 'Canada',
    'IN': 'India',
    'DE': 'Germany',
    'CH': 'Switzerland',
    'AU': 'Australia',
    'FR': 'France',
    'NL': 'Netherlands',
    'SE': 'Sweden',
    'CN': 'China',
    'FI': 'Finland',
    'SG': 'Singapore',
    'DK': 'Denmark',
    'IE': 'Ireland',
    'IT': 'Italy',
    'JP': 'Japan',
    'ES': 'Spain',
    'IL': 'Israel',
    'MX': 'Mexico',
    'AT': 'Austria',
    'KR': 'South Korea',
    'AE': 'United Arab Emirates',
    'BE': 'Belgium',
    'BM': 'Bermuda',
    'HK': 'Hong Kong',
    'TW': 'Taiwan',
    'NZ': 'New Zealand',
    'SA': 'Saudi Arabia',
    'AR': 'Argentina',
    'NO': 'Norway',
    'HU': 'Hungary',
    'UY': 'Uruguay',
    'ID': 'Indonesia',
    'VN': 'Vietnam',
    'KY': 'Cayman Islands',
    'HR': 'Croatia',
    'TR': 'Turkey',
    'BR': 'Brazil',
    'CZ': 'Czech Republic',
    'GE': 'Georgia',
    'MD': 'Moldova',
    'IS': 'Iceland',
    'PR': 'Puerto Rico',
    'CO': 'Colombia',
    'KE': 'Kenya',
    'LU': 'Luxembourg',
    'PH': 'Philippines'
}

# Applying the mapping to the 'country' column to replace ISO codes with full country names
companies_names['country_full_name'] = companies_names['country'].map(iso_to_full_country).fillna(companies_names['country'])

# Display the dataset to verify the transformation 
print(companies_names[['company_id', 'country', 'country_full_name']].head())


   company_id country country_full_name
0        1009      US     United States
1        1016      US     United States
2        1021      US     United States
3        1025      US     United States
4        1028      US     United States


In this code:

I created a dictionary, iso_to_full_country, where the keys are the provided ISO codes and the values are the full country names.

I then used the map() function on the country column to replace the ISO codes with the full country names. If an ISO code wasn't found in the dictionary, the original value from the country column was retained using fillna(df_companies['country'])

In [133]:
# Counting occurrences of each country in the new country_full_name column
country_counts3 = companies_names['country_full_name'].value_counts()

In [134]:
#printing results
print(country_counts3)

country_full_name
United States           5419
0                        156
United Kingdom           129
Canada                    71
India                     39
Germany                   35
Switzerland               29
Australia                 19
France                    19
Netherlands               15
Sweden                    11
China                     11
Finland                    9
Singapore                  8
Denmark                    8
Ireland                    7
Italy                      7
OO                         6
Japan                      5
Spain                      5
Israel                     4
Mexico                     4
Austria                    4
South Korea                3
United Arab Emirates       3
Belgium                    3
Bermuda                    3
Hong Kong                  3
Taiwan                     3
New Zealand                2
Saudi Arabia               2
Argentina                  2
Norway                     2
Hungary                  

In [137]:
# Marking '00' and '0' values in the 'country_full_name' column as 'Unknown'
companies_names['country_full_name'] = companies_names['country_full_name'].replace({'00': 'Unknown', '0': 'Unknown'})


Data Imputation

In [135]:
# Counting occurrences of each city 
city_counts2 = companies_names['city'].value_counts()

#print results
print(city_counts2)



city
New York         275
0                219
Chicago          128
San Francisco    107
Houston          107
                ... 
Jiangsu            1
Coplay             1
Oviedo             1
Verona             1
Tewksbury          1
Name: count, Length: 1675, dtype: int64


In [138]:
# Replacing '0' values in the 'city' column with 'Unknown'
companies_names['city'] = companies_names['city'].replace('0', 'Unknown')


In [140]:
# Counting occurrences of each city 
city_counts3 = companies_names['city'].value_counts()

#print results
print(city_counts3)

city
New York         275
Unknown          219
Chicago          128
San Francisco    107
Houston          107
                ... 
Jiangsu            1
Coplay             1
Oviedo             1
Verona             1
Tewksbury          1
Name: count, Length: 1675, dtype: int64


In [142]:
# Checking the unique values and their counts for the 'state' column
state_counts2 = companies_names['state'].value_counts()

# Displaying the results
print(state_counts2.head(20))  # Displaying top 20 states by count for a concise view


state
0             480
CA            470
California    465
Texas         322
NY            219
Florida       165
New York      156
TX            146
New Jersey    140
Illinois      132
IL            123
FL            112
PA            106
MA            106
Georgia       102
Virginia       96
VA             88
Ohio           87
GA             87
NJ             81
Name: count, dtype: int64


Iconsistencies in Naming: There are inconsistencies in state naming. For instance, "California" and "CA" both represent the state of California. Similarly, "Texas" and "TX" or "New York" and "NY" are other examples.

Placeholders: There's a placeholder or missing value represented as "0" with 480 occurrences in the dataset. 

In [144]:


# Mapping of state abbreviations to full state names
state_abbreviations2 = {
    'AL': 'Alabama', 'AK': 'Alaska', 'AZ': 'Arizona', 'AR': 'Arkansas', 'CA': 'California', 
    'CO': 'Colorado', 'CT': 'Connecticut', 'DE': 'Delaware', 'FL': 'Florida', 'GA': 'Georgia', 
    'HI': 'Hawaii', 'ID': 'Idaho', 'IL': 'Illinois', 'IN': 'Indiana', 'IA': 'Iowa', 
    'KS': 'Kansas', 'KY': 'Kentucky', 'LA': 'Louisiana', 'ME': 'Maine', 'MD': 'Maryland', 
    'MA': 'Massachusetts', 'MI': 'Michigan', 'MN': 'Minnesota', 'MS': 'Mississippi', 'MO': 'Missouri', 
    'MT': 'Montana', 'NE': 'Nebraska', 'NV': 'Nevada', 'NH': 'New Hampshire', 'NJ': 'New Jersey', 
    'NM': 'New Mexico', 'NY': 'New York', 'NC': 'North Carolina', 'ND': 'North Dakota', 'OH': 'Ohio', 
    'OK': 'Oklahoma', 'OR': 'Oregon', 'PA': 'Pennsylvania', 'RI': 'Rhode Island', 'SC': 'South Carolina', 
    'SD': 'South Dakota', 'TN': 'Tennessee', 'TX': 'Texas', 'UT': 'Utah', 'VT': 'Vermont', 
    'VA': 'Virginia', 'WA': 'Washington', 'WV': 'West Virginia', 'WI': 'Wisconsin', 'WY': 'Wyoming'
}

# Apply the mapping to the 'state' column to replace abbreviations with full state names
companies_names['state_full_name'] = companies_names['state'].map(state_abbreviations2).fillna(companies_names['state'])

# printing changes
print(companies_names[['company_id', 'state', 'state_full_name']].head(20))


    company_id      state state_full_name
0         1009         NY        New York
1         1016          0               0
2         1021         NY        New York
3         1025      Texas           Texas
4         1028      Texas           Texas
5         1038          0               0
6         1043          0               0
7         1044          0               0
8         1060  Stockholm       Stockholm
9         1063         CA      California
10        1068         NY        New York
11        1103         NJ      New Jersey
12        1115         BW              BW
13        1116       Ohio            Ohio
14        1123         NC  North Carolina
15        1185   New York        New York
16        1241          0               0
17        1248     London          London
18        1254         CA      California
19        1259          0               0


In [145]:
# Replacing '0' values in the 'state_full_name' column with 'Unknown'
companies_names['state_full_name'] = companies_names['state_full_name'].replace('0', 'Unknown')


In [146]:
# printing changes
print(companies_names[['company_id', 'state', 'state_full_name']].head(20))

    company_id      state state_full_name
0         1009         NY        New York
1         1016          0         Unknown
2         1021         NY        New York
3         1025      Texas           Texas
4         1028      Texas           Texas
5         1038          0         Unknown
6         1043          0         Unknown
7         1044          0         Unknown
8         1060  Stockholm       Stockholm
9         1063         CA      California
10        1068         NY        New York
11        1103         NJ      New Jersey
12        1115         BW              BW
13        1116       Ohio            Ohio
14        1123         NC  North Carolina
15        1185   New York        New York
16        1241          0         Unknown
17        1248     London          London
18        1254         CA      California
19        1259          0         Unknown


In [147]:
# checking columns
companies_names.dtypes

company_id             int64
name                  object
description           object
company_size         float64
state                 object
country               object
city                  object
zip_code              object
address               object
url                   object
country_full_name     object
state_full_name       object
dtype: object

### Dropping unnecessary columns for analysis

In [148]:
# Dropping the 'zip_code', 'address', and 'url' columns from the dataframe
companies_names = companies_names.drop(columns=['zip_code', 'address', 'url','description'])


In [149]:
# checking dataframe
companies_names.head()

company_id                        name  company_size  state country  \
0        1009                         IBM           7.0     NY      US   
1        1016               GE HealthCare           7.0      0      US   
2        1021                    GE Power           7.0     NY      US   
3        1025  Hewlett Packard Enterprise           7.0  Texas      US   
4        1028                      Oracle           7.0  Texas      US   

               city country_full_name state_full_name  
0  Armonk, New York     United States        New York  
1           Chicago     United States         Unknown  
2       Schenectady     United States        New York  
3           Houston     United States           Texas  
4            Austin     United States           Texas

In [150]:
# Renaming the specified columns
companies_names = companies_names.rename(columns={
    'country_full_name': 'company_location_country',
    'city': 'company_location_city',
    'state_full_name': 'company_location_state',
    'name': 'company_name'
})



In [151]:
# checking dataframe
companies_names.head()

company_id                company_name  company_size  state country  \
0        1009                         IBM           7.0     NY      US   
1        1016               GE HealthCare           7.0      0      US   
2        1021                    GE Power           7.0     NY      US   
3        1025  Hewlett Packard Enterprise           7.0  Texas      US   
4        1028                      Oracle           7.0  Texas      US   

  company_location_city company_location_country company_location_state  
0      Armonk, New York            United States               New York  
1               Chicago            United States                Unknown  
2           Schenectady            United States               New York  
3               Houston            United States                  Texas  
4                Austin            United States                  Texas

In [153]:
# Dropping additional columns
companies_names = companies_names.drop(columns=['country', 'state'])


In [154]:
# checking columns
companies_names.dtypes

company_id                    int64
company_name                 object
company_size                float64
company_location_city        object
company_location_country     object
company_location_state       object
dtype: object

### Conducting additional data consistency checks
It seems that some cities have additional location information (like the state or region) appended after the comma, while others might have trailing commas or other inconsistent patterns.

In [156]:
# Counting the number of cities that include a comma
cities_with_comma2 = companies_names[companies_names['company_location_city'].str.contains(',', na=False)]
number_of_cities_with_comma2 = len(cities_with_comma2)

# Displaying the number and a sample of cities with a comma
cities_with_comma_sample2 = cities_with_comma2['company_location_city'].sample(min(20, number_of_cities_with_comma2))

print(f"Number of cities with a comma: {number_of_cities_with_comma2}")
print("\nSample of cities with a comma:")
print(cities_with_comma_sample2)


Number of cities with a comma: 33

Sample of cities with a comma:
4333                            Plainsboro, United States
3402                                       Washington, DC
3414    Denmark, Germany, UK, USA, Canada, Australia, ...
3143                            Florham Park, New Jersey 
3200                          Kohala Coast, Hawaii Island
4286                                          Atlanta, GA
4872                              San Francisco, Bay area
1836                                   St Louis, MO 63141
138            Urbana-Champaign, Chicago, and Springfield
3036                                 Hartland, Village of
2285                                       Washington, DC
3530                                        La Grange, IL
4066                                       Chuo-ku, Osaka
2506                                          Alpharetta,
5093                                           Dublin, OH
2127                                       Israel, Afula 
96    

In [157]:
# Removing states or regions from the 'company_location_city' column by keeping only the part before the comma
companies_names['company_location_city'] = companies_names['company_location_city'].str.split(',').str[0].str.strip()


In [158]:
#checking results
companies_names.head()

company_id                company_name  company_size company_location_city  \
0        1009                         IBM           7.0                Armonk   
1        1016               GE HealthCare           7.0               Chicago   
2        1021                    GE Power           7.0           Schenectady   
3        1025  Hewlett Packard Enterprise           7.0               Houston   
4        1028                      Oracle           7.0                Austin   

  company_location_country company_location_state  
0            United States               New York  
1            United States                Unknown  
2            United States               New York  
3            United States                  Texas  
4            United States                  Texas

In [159]:
# exporting dataframe to csv.
companies_names.to_csv(os.path.join(path, 'DATA','CLEAN DATA', 'company_name.csv'))

### creating the data profile

In [162]:

# Creating a data profile for the dataset
data_profile2 = {
    'Variables': companies_names.columns,
    'Data Types': companies_names.dtypes,
    'Description': ['' for _ in range(companies_names.shape[1])],  
    'Time-variant/-invariant': ['' for _ in range(companies_names.shape[1])],
    'Structured/Unstructured': ['Structured' if dtype != 'object' else '' for dtype in companies_names.dtypes],
    'Qualitative/Quantitative': ['Quantitative' if dtype in ['int64', 'float64'] else 'Qualitative' for dtype in companies_names.dtypes],
    'Qualitative: Nominal/Ordinal': ['Nominal' if dtype == 'object' else '' for dtype in companies_names.dtypes],
    'Quantitative: Discrete/Continuous': ['Discrete' if dtype == 'int64' else 'Continuous' if dtype == 'float64' else '' for dtype in companies_names.dtypes]
}

# Converting the data profile to a DataFrame for display
data_profile2 = pd.DataFrame(data_profile2)

# Displaying the data profile (optional in your Jupyter Notebook)
print(data_profile2)


                                         Variables Data Types Description  \
company_id                              company_id      int64               
company_name                          company_name     object               
company_size                          company_size    float64               
company_location_city        company_location_city     object               
company_location_country  company_location_country     object               
company_location_state      company_location_state     object               

                         Time-variant/-invariant Structured/Unstructured  \
company_id                                                    Structured   
company_name                                                               
company_size                                                  Structured   
company_location_city                                                      
company_location_country                                                   
comp

In [163]:
# exporting dataframe to csv.
data_profile2.to_csv(os.path.join(path, 'DATA','Data profile', 'data_profile_companies_names.csv'))

## Importing the 4th dataframe (job_skills)

In [165]:
job_skills = pd.read_csv(r'/Users/pierremora/Desktop/LinkedIn Job Postings 2023/DATA/RAW DATA/job_skills.csv', index_col = False)

### Descriptive analysis


In [166]:
job_skills.describe()

job_id
count  2.789900e+04
mean   3.692216e+09
std    9.702447e+07
min    8.500877e+07
25%    3.693072e+09
50%    3.697358e+09
75%    3.699413e+09
max    3.701374e+09

In [167]:
job_skills.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27899 entries, 0 to 27898
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   job_id     27899 non-null  int64 
 1   skill_abr  27899 non-null  object
dtypes: int64(1), object(1)
memory usage: 436.1+ KB


In [168]:
# checking job_skills data types
job_skills.dtypes

job_id        int64
skill_abr    object
dtype: object

In [169]:
#checking head dataframe
job_skills.head()

job_id skill_abr
0  3690843087      ACCT
1  3690843087       FIN
2  3691763971      MGMT
3  3691763971      MNFC
4  3691775263      MGMT

### Checking missing values

In [170]:
# Checking for missing values and their respective percentages
missing_values5 = job_skills.isnull().sum()
missing_percentage5 = (missing_values5 / len(job_skills)) * 100

missing_data5 = pd.DataFrame({
    'Missing Values': missing_values5,
    'Percentage (%)': missing_percentage5
})

print(missing_data5)


           Missing Values  Percentage (%)
job_id                  0             0.0
skill_abr               0             0.0


There are no missing values in the "job_skills.csv" dataset. Both the job_id and skill_abr columns have complete data.

### Checking for Duplicates or fully duplicates

In [171]:
# Checking for duplicate rows
duplicates5 = job_skills.duplicated().sum()

# Check for duplicate job_id values (to see if there are jobs with multiple skills)
duplicate_job_ids5 = job_skills.duplicated(subset=['job_id']).sum()

print(f"Number of fully duplicated rows: {duplicates5}")
print(f"Number of duplicated job_id values: {duplicate_job_ids5}")


Number of fully duplicated rows: 0
Number of duplicated job_id values: 12262


There are 0 fully duplicated rows in the dataset. This means there are no two rows that are identical in both job_id and skill_abr.

There are 12,262 duplicated job_id values. This indicates that there are jobs associated with multiple skills (which is expected in a job skills dataset).

### data consistency

In [172]:
# Inspecting  unique values of the 'job_id' column
unique_job_ids5 = job_skills['job_id'].unique()
sample_job_ids5 = unique_job_ids5[:20]  # Displaying first 20 for brevity
print("Sample of unique job_ids:")
print(sample_job_ids5)
print("\n")

# Inspecting unique values of the 'skill_abr' column
unique_skill_abr5 = job_skills['skill_abr'].unique()
sample_skill_abr5 = unique_skill_abr5[:20]  # Displaying first 20 for brevity
print("Sample of unique skill abbreviations:")
print(sample_skill_abr5)


Sample of unique job_ids:
[3690843087 3691763971 3691775263 3691786992 3691789797 3691789919
 3691792844 3691793575 3691794313 3691795305 3691795389 3691795980
 3691796226 3691797249 3691797924 3691797973 3691797979 3691797980
 3691797982 3691798875]


Sample of unique skill abbreviations:
['ACCT' 'FIN' 'MGMT' 'MNFC' 'HCPR' 'ENG' 'IT' 'ADM' 'SALE' 'DSGN' 'ART'
 'EDU' 'TRNG' 'BD' 'PRJM' 'CNSL' 'STRA' 'OTHR' 'RSCH' 'GENB']


### dataset is clean and consistent

In [173]:
#exporting dataframe to csv.

job_skills.to_csv(os.path.join(path, 'DATA','CLEAN DATA', 'job_skills_clean.csv'))

### creating data profile


In [175]:

# List to store the data profile information
data_profile_list5 = []

# Loop through each column to gather the data profile details
for col in job_skills.columns:
    dtype = job_skills[col].dtype
    structured = 'Structured' if dtype != 'object' else 'Unstructured'
    qual_quant = 'Quantitative' if dtype in ['int64', 'float64'] else 'Qualitative'
    nom_ord = 'Nominal' if dtype == 'object' else ''
    disc_cont = 'Discrete' if dtype == 'int64' else 'Continuous' if dtype == 'float64' else ''
    
    data_profile_list5.append([col, dtype, '', '', structured, qual_quant, nom_ord, disc_cont])

# Converting the list to a DataFrame
columns = ['Variables', 'Data Types', 'Description', 'Time-variant/-invariant', 'Structured/Unstructured', 'Qualitative/Quantitative', 'Qualitative: Nominal/Ordinal', 'Quantitative: Discrete/Continuous']
data_profile_skills_df = pd.DataFrame(data_profile_list5, columns=columns)

# Display the data profile (optional in your Jupyter Notebook)
print(data_profile_skills_df)


   Variables Data Types Description Time-variant/-invariant  \
0     job_id      int64                                       
1  skill_abr     object                                       

  Structured/Unstructured Qualitative/Quantitative  \
0              Structured             Quantitative   
1            Unstructured              Qualitative   

  Qualitative: Nominal/Ordinal Quantitative: Discrete/Continuous  
0                                                       Discrete  
1                      Nominal                                    


In [177]:
#exporting dataframe to csv.

data_profile_skills_df.to_csv(os.path.join(path, 'DATA','Data profile', 'job_skills_data_profile.csv'))